### PR TITLE
Hooke-Jeeves  Method

### PR DESCRIPTION
**Pull request by VladislavSevruk**

### CODE REVIEW

**Comment by PavelKovalets** [comment]
- Add @AlexanderSoroka to the list of reviewers.
- include all [pull request changes as a single commit](http://ndlib.github.io/practices/one-commit-per-pull-request/)

**Comment by VladislavSevruk** [comment]
- Only rep owner can do this, I guess.
- Done.

File: include/k52/optimization/hooke_jeeves_method.h [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+#include <boost/shared_ptr.hpp>
+
+#include <k52/optimization/continuous_optimizer.h>
+
+namespace k52
+{
+namespace optimization
+{
+
+class HookeJeevesMethod: public ContinuousOptimizer
```
Comment (AlexanderSoroka) [inline include/k52/optimization/hooke_jeeves_method.h:22]: Add comment with the link to implementation description

File: include/k52/optimization/hooke_jeeves_method.h [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+{
+public:
+	HookeJeevesMethod(double precision = 1e-10,
```
Comment (AlexanderSoroka) [inline include/k52/optimization/hooke_jeeves_method.h]: add doxygen comment with method arguments description

File: include/k52/optimization/hooke_jeeves_method.h [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+#ifdef BUILD_WITH_MPI
+	virtual void Send(boost::mpi::communicator* communicator, int target) const;
+
+	virtual void Receive(boost::mpi::communicator* communicator, int source);
+#endif
+
+protected:
+	virtual std::vector<double> FindOptimalParameters(const std::vector<double>&);
+
+	/// Checks if steps vector length
```
Comment (AlexanderSoroka) [inline include/k52/optimization/hooke_jeeves_method.h]: Use doxygen tags in comment (@brief, @param etc)

File: src/optimization/hooke_jeeves_method.cpp [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+#include <algorithm>
+#include <stdexcept>
+
+#include <k52/optimization/params/i_continuous_parameters.h>
+
+using ::std::vector;
+
+namespace
+{
+	double kAcceleration = 3.5;
```
Comment (AlexanderSoroka) [inline src/optimization/hooke_jeeves_method.cpp]: Should it be const? Why 3.5?

File: src/optimization/hooke_jeeves_method.cpp [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+}
+
+namespace k52
+{
+namespace optimization
+{
+
+HookeJeevesMethod::HookeJeevesMethod(double precision, size_t max_iteration_number,
+	double init_step)
+	: precision_(precision)
```
Comment (AlexanderSoroka) [inline src/optimization/hooke_jeeves_method.cpp]: wrong initialization order

File: src/optimization/hooke_jeeves_method.cpp [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+	, max_iteration_number_(max_iteration_number)
+	, init_step_(init_step)
+{
+}
+
+HookeJeevesMethod* HookeJeevesMethod::Clone() const
+{
+	return new HookeJeevesMethod();
```
Comment (AlexanderSoroka) [inline src/optimization/hooke_jeeves_method.cpp]: Wrong implementation. The members of clone must be initialized with the same values as in *this

File: src/optimization/hooke_jeeves_method.cpp [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+	vector<double> arguments_after_search(initial_parameters.size());
+	vector<double> steps_array = vector<double>(arguments.size(), init_step_);
+	size_t iteration = 0;
+
+	do
+	{
+		arguments_after_search = CoordinatewiseSearch(arguments, steps_array);
+		if (arguments == arguments_after_search)
+		{
+			for (size_t i = 0; i < steps_array.size(); i++)
```
Comment (AlexanderSoroka) [inline src/optimization/hooke_jeeves_method.cpp]: It doesn't play a big role in modern c++ but prefix increment (++i) form is still preferable

File: include/k52/optimization/hooke_jeeves_method.h [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+	/// satisfies an exit criteria
+	bool IsExitCriteriaFulfilled(std::vector<double>);
```
Comment (AlexanderSoroka) [inline include/k52/optimization/hooke_jeeves_method.h]: Argument must be passed by const reference not by value

File: include/k52/optimization/hooke_jeeves_method.h [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+	/// Realization of Coordinate wise search,
+	/// Hooke-Jeeves is based on this optimization method
+	std::vector<double> CoordinatewiseSearch(std::vector<double>, std::vector<double>);
```
Comment (AlexanderSoroka) [inline include/k52/optimization/hooke_jeeves_method.h]: Arguments must be passed by const reference not by value

File: include/k52/optimization/hooke_jeeves_method.h (hunk 8 above)
Comment (AlexanderSoroka) [inline include/k52/optimization/hooke_jeeves_method.h]: It seems to be const method

File: include/k52/optimization/hooke_jeeves_method.h (hunk 9 above)
Comment (AlexanderSoroka) [inline include/k52/optimization/hooke_jeeves_method.h]: It seems to be const method

File: include/k52/optimization/hooke_jeeves_method.h [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+
+private:
+
+	/// Maximum iterations number
```
Comment (AlexanderSoroka) [inline include/k52/optimization/hooke_jeeves_method.h]: It's better to keep members alphabetically ordered

File: src/optimization/hooke_jeeves_method.cpp [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+			{
+				steps_array[i] /= 10;
+			}
+			continue;
+		}
+
+		vector<double> next_step_arguments (arguments.size())
```
Comment (AlexanderSoroka) [inline src/optimization/hooke_jeeves_method.cpp]: Variables must be defined separately each on new line

File: src/optimization/hooke_jeeves_method.cpp (hunk 4 above)
Comment (VladislavSevruk) [inline src/optimization/hooke_jeeves_method.cpp]: Yep, researches, that was in our Methods of computational experiment course, show that minimum number of iterations will be achieved by coefficient that equals 3.5. Testing different coefficient values by continuous_optimization_examples confirms this.

File: src/optimization/hooke_jeeves_method.cpp (hunk 5 above)
Comment (VladislavSevruk) [inline src/optimization/hooke_jeeves_method.cpp]: I'm not sure that I've got the point. Is it so critical? Should I initialize variables in alphabetical order or some other way?

Comment (AlexanderSoroka) [inline src/optimization/hooke_jeeves_method.cpp]: The initialization order should be the same as declaration order.

Comment (VladislavSevruk) [inline src/optimization/hooke_jeeves_method.cpp]: Clear.

**Comment by VladislavSevruk** [comment]
@AlexanderSoroka, 
Take a look at this, please.

File: include/k52/optimization/hooke_jeeves_method.h [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+	virtual void Receive(boost::mpi::communicator* communicator, int source);
+#endif
+
+protected:
+	virtual std::vector<double> FindOptimalParameters(const std::vector<double>&);
+
+	/// @brief Checks if steps vector length satisfies an exit criteria.
+	/// @param steps_array - vector of steps values for diffirent coordinates 
+	/// @return true if steps vector norm less than specified precision, false other way
+	const bool IsExitCriteriaFulfilled(
```
Comment (AlexanderSoroka) [inline include/k52/optimization/hooke_jeeves_method.h]: bool IsExitCriteriaFulfilled() const

File: include/k52/optimization/hooke_jeeves_method.h [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+		const std::vector<double>& steps_array);
+
+	/// @brief Realization of Coordinate wise search.
+	/// Method implements search from start point in all directions with specified step value.
+	/// Hooke-Jeeves method is based on this optimization method.
+	/// @param arguments - vector of coordinates of start point
+	/// @param steps_array - vector of values of step for different coordinates
+	/// @return vector of new coordinates
+	const std::vector<double> CoordinatewiseSearch(
```
Comment (AlexanderSoroka) [inline include/k52/optimization/hooke_jeeves_method.h]: @see line 46

**Comment by VladislavSevruk** [comment]
@AlexanderSoroka, 
One more try.

File: src/optimization/hooke_jeeves_method.cpp (hunk 4 above)
Comment (AlexanderSoroka) [inline src/optimization/hooke_jeeves_method.cpp]: const double?

Comment (VladislavSevruk) [inline src/optimization/hooke_jeeves_method.cpp]: Well, I suppose to use exactly this value every time but I can implement this value as class field and initialize it in constructor, for example. Should I realize it in this way?

Comment (AlexanderSoroka) [inline src/optimization/hooke_jeeves_method.cpp]: Yes, it's better to make it configurable

Comment (VladislavSevruk) [inline src/optimization/hooke_jeeves_method.cpp]: Done.

**Comment by VladislavSevruk** [comment]
@AlexanderSoroka,
anything else or it's OK for now?

**Comment by AlexanderSoroka** [comment]
Looks good for me

File: include/k52/optimization/hooke_jeeves_method.h [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+/// @brief Realization of Hooke-Jeeves optimization method.
+class HookeJeevesMethod: public ContinuousOptimizer
+{
+public:
+	/// @brief Creates instance of HookeJeevesMethod.
+	/// @param init_step - number that defines value of step on first iteration
+	/// @param max_iteration_number - maximum number of iterations before computations will be stopped
+	/// @param precision - number that defines accuracy of our computations
+	HookeJeevesMethod(
+		double acceleration = 3.5,
```
Comment (PavelKovalets) [inline include/k52/optimization/hooke_jeeves_method.h]: Missing comment with detailed description of parameter with explanation why default value is better

File: include/k52/optimization/hooke_jeeves_method.h [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+#include <boost/shared_ptr.hpp>
+
+#include <k52/optimization/continuous_optimizer.h>
+
+namespace k52
+{
+namespace optimization
+{
+
+/// @class HookeJeevesMethod
```
Comment (PavelKovalets) [inline include/k52/optimization/hooke_jeeves_method.h:20]: Add link to the source of your algorithm
It hugely differs from the one described in [WiKi](https://ru.wikipedia.org/wiki/%D0%9C%D0%B5%D1%82%D0%BE%D0%B4_%D0%A5%D1%83%D0%BA%D0%B0_%E2%80%94_%D0%94%D0%B6%D0%B8%D0%B2%D1%81%D0%B0)

File: src/optimization/hooke_jeeves_method.cpp [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+vector<double> HookeJeevesMethod::CoordinatewiseSearch(
+	const vector<double>& arguments,
+	const vector<double>& steps_array) const
+{
+	vector<double> new_arguments(arguments);
+	vector<double> temp(arguments);
+	for (size_t i = 0; i < arguments.size(); ++i)
+	{
+		temp[i] = new_arguments[i] + steps_array[i];
+		if (CountObjectiveFunctionValueToMinimize(temp) >= CountObjectiveFunctionValueToMinimize(new_arguments))
```
Comment (PavelKovalets) [inline src/optimization/hooke_jeeves_method.cpp]: CountObjectiveFunctionValueToMinimize may be reeeealy time consuming (hours to calculate).
So must not recalculate it with same arguments (e.g.  new_arguments not changed) but save the result in variable instead

File: src/optimization/hooke_jeeves_method.cpp [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+{
+	vector<double> arguments(initial_parameters);
+	vector<double> arguments_after_search(initial_parameters.size());
+	vector<double> steps_array = vector<double>(arguments.size(), init_step_);
+	size_t iteration = 0;
+
+	do
+	{
+		arguments_after_search = CoordinatewiseSearch(arguments, steps_array);
+		if (arguments == arguments_after_search)
```
Comment (PavelKovalets) [inline src/optimization/hooke_jeeves_method.cpp]: Invalid comparison of floating point values. Potentially an endless loop.
See k52.common.FloatingPoint. You need to add corresponding method for vector and reuse existing AreEqual there, also add corresponding unit tests.

File: src/optimization/hooke_jeeves_method.cpp [hunk 18]
```diff
@@ -0,0 +1,157 @@
...
+		{
+			for (size_t i = 0; i < steps_array.size(); ++i)
+			{
+				steps_array[i] /= 10;
```
Comment (PavelKovalets) [inline src/optimization/hooke_jeeves_method.cpp]: Why 10? Also convert to field with default value

**Comment by PavelKovalets** [comment]
- Use 4 spaces instead of tabs everywhere
- Fix my inline comments (see files list)

File: include/k52/optimization/hooke_jeeves_method.h [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+public:
+	/// @brief Creates instance of HookeJeevesMethod.
+	/// @param init_step - number that defines value of step on first iteration
+	/// @param max_iteration_number - maximum number of iterations before computations will be stopped
+	/// @param precision - number that defines accuracy of our computations
+	HookeJeevesMethod(
+		double acceleration = 3.5,
+		double init_step = 1,
+		size_t max_iteration_number = 10000,
+		double precision = 1e-10);
```
Comment (PavelKovalets) [inline include/k52/optimization/hooke_jeeves_method.h]: use Constants::Eps

File: src/optimization/hooke_jeeves_method.cpp [hunk 20]
```diff
@@ -0,0 +1,161 @@
...
+    for (size_t i = 0; i < arguments.size(); ++i)
+    {
+        double new_function_value = CountObjectiveFunctionValueToMinimize(new_arguments);
+        temp[i] = new_arguments[i] + steps_array[i];
+        if (CountObjectiveFunctionValueToMinimize(temp) >= new_function_value)
+        {
+            temp[i] = new_arguments[i] - steps_array[i];
+            if (CountObjectiveFunctionValueToMinimize(temp) >= new_function_value)
+            {
+                temp[i] = new_arguments[i];
```
Comment (PavelKovalets) [inline src/optimization/hooke_jeeves_method.cpp]: Not very clear logic with temp and new_arguments values, which are sometimes reset to their initial values and sometimes new values are applied. This should be just 2 tests for positive and negative step, this should be more clear. 
Try to refactor to achieve more readable code - it should be self-explanatory. Do not afraid to copy whole vectors where necessary (not reset sole indexes) if necessary - just assign one vector variable to another then.

**Comment by PavelKovalets** [comment]
Everything look really good. However there are a few comment, but not very significant and should be easy to fix
- refactor CoordinatewiseSearch to be more readable (see inline comment)
- still some tabs present. Try to remove them (search and replace in text editor for example)
- Cyrillic symbols are forbidden even in comments. Please just transliterate then in you links to sources (its ok that the actual file name is different, we'll still be able to find them)

**Comment by PavelKovalets** [comment]
And also please post a message here when you'll fix this. I do not received the notification on code change for some reason

**Comment by VladislavSevruk** [comment]
@PavelKovalets ,
I hope it's the final one.

**Comment by VladislavSevruk** [comment]
@PavelKovalets ,
Ooops, I found up that I forgot to add step_divider_ parameter into Clone and BUILD_WITH_MPI methods. I think you better to know about it.

**Comment by PavelKovalets** [comment]
I've fixed this [during merge](https://github.com/PavelKovalets/k52/commit/57c256229c9c056c9f505a020098589c744d85bd) )

**Pull request merged**